accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -616,6 +609,442 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+type or its nullable variant, or a value of type `Null`, on all paths; and the
+variable is treated as having nullable type.
+
+
+```dart
+
+
+```
+```dart
+void test(bool b) {
+  { // late and/or final the same
+    var x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    } else {
+      x = 3.0;
+    }
+    use(x) // Error: x has no inferred type
+    x = 3; // Error: x has no inferred type
+  }
+
+{ // late and/or final the same
+    var x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    } else {
+      x = null;
+    } // x has type int?
+  }
+
+  { // late and/or final the same
+    var x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    } else {
+      x = null as int?;
+    } // x has type int?
+  }
+
+  {
+    var x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    }
+    // x has type int?
+  }
+
+  { // late final the same
+    late x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    }
+    // x has type int
+  }
+
+  {
+    final x; // x is an untyped variable
+    if (b) {
+      x = 3;
+    }
+    // Error: not assigned on all paths
+  }
+
+  { // late and/or final the same
+    var x; // x is an untyped variable
+    if (b) {
+      if (b) {
+        x = 3;
+      } else {
+        x = 3;
+      }
+      x.isEven; // No error
+    } else {
+      x = "hello";
+      x.length; // No error
+    } // Error: x has no consistent inferred type
+  }
+
+  { // late and/or final the same
+    var x; // x is an untyped variable
+
+    if (b) {
+      if (b) {
+        x = 3;
+        return;
+      } else {
+        return;
+      } // x is inferred to int
+    } else {
+      x = 3;
+    } // x is inferred to int
+    x.isEven; // No error
+  }
+
+  { // late and/or final the same
+    var x; // x is an untyped variable
+
+    if (b) {
+      if (b) {
+        x = 3;
+        return;
+      } else {
+        x = null;
+        return;
+      } // x is inferred as int?
+    } else {
+      x = 3;
+    } // x is inferred as int?
+  }
+
+  {
+    var z; // z is an untyped variable
+    if (b) {
+      z = 3; // z is promoted to int
+     } else {
+      z = 3.0; // z is promoted to double
+      return;
+    } // Error, z has no consistent inferred type
+  }
+
+  {
+    var z; // z is an untyped variable
+    if (b) {
+      z = 3; // z is promoted to int
+      return;
+     } else {
+      z = 3.0; // z is promoted to double
+      return;
+    } // Error, z has no consistent inferred type
+  }
+
+  {
+    var u; // u is an untyped variable
+    if (b) {
+      u = 0 as num; // u is promoted to num
+      if (u is! int) u = 1;
+      // u has inferred type num, but is promoted to int
+    } else {
+      u = 3;
+      // u is inferred as int
+    } Error: u has no inferred type
+  }
+}
+```
+
+It is an error to read an **untyped variable** when it has no **initialization
+inferred type**.
+
+It is an error to write to an **untyped variable** when it has no
```

Comment:
Do we need this?  In all the cases I can think of where an untyped variable is written to and there's no initialization inferred type, either the write is an initializing assignment or there is already some other error.

Reply:
It depends on the exact details of the proposal, but for this variant at least, yes:

var x;
if (true) {
  while(b) {
    x = 3; // Not an initializing write
  }
  // x is still untyped
} else {
  x = "hello"; // x inferred to String
} // x is untyped on one branch, `String` on another, so infer `String?`
x?.length; // x is a non-null int, what does this mean?
```